api/src/app.ts

```diff
@@ -0,0 +1,193 @@
+ *
+ * As presently written, versions can only be whole integers.
+ *
+ */
+const apiVersionPath = /^v(\d+)$/;
+
+/**
+ * @todo Set using environment variables for security hardening.
+ */
+app.use(function (req, res, next) {
+  res.setHeader("Access-Control-Allow-Origin", "*");
+  res.setHeader("Access-Control-Allow-Methods", "*");
+  res.setHeader("Access-Control-Allow-Headers", "*");
+  next();
+});
+
+const awaitErrorHandlerFactory = (middleware: any) => {
+  return async (request: Request, response: Response, next: NextFunction) => {
+    try {
+      await middleware(request, response, next);
+    } catch (err) {
+      console.error(err);
+      next(err);
+    }
+  };
+};
+
+const versionMessages = {
+  200: {
+    title: "OK",
+    detail: "Everything is fine.",
+  },
+  400: {
+    title: "No API version specified",
+    detail: `An API version must be specified, but no version was specified.
+        To specify a version, use one of the following approaches:
+        (1) Specify the version in the first segment of the path, e.g. \`/v1/grants\`
+        (2) Specify the version in the Accept header: \`Accept: application/vnd.api+json; gov.usda.fs.ga.api.version=1\``,
+  },
+  404: {
+    title: "Version not found",
+    detail: `The version specified is not an available API version. Please choose from the following available versions: ${apiVersions}`,
+  },
+  409: {
+    title: "Conflicting versions requested",
+    detail:
+      "The Accept header and the request path each requested different, conflicting versions. Please ensure that only one version is specified. You can set both the path and header to the same version, or you can request a version using only the path or only the header.",
+  },
+};
+
+/**
+ * Adds a `version` property to Express.Request, so we can attach
+ * that metadata as we resolve the version.
+ */
+declare global {
+  namespace Express {
+    interface Request {
+      version: number;
+    }
+  }
+}
+
+app.use(function (request, response, next) {
+  let [statusCode, version] = resolveVersions(request);
+  if (statusCode == 200) {
+    /**
+     * If the first segment includes the version,
+     * rewrite the path without the version.
+     */
+    if (request.url.match(/^\/v\d+/)) {
+      request.url = `/${request.url.split("/").slice(2).join("/")}`;
+    }
+    request.version = version;
+    next();
+  } else {
+    return response.status(statusCode).json({
+      errors: [
+        {
+          status: statusCode,
+          title: versionMessages[statusCode].title,
+          detail: versionMessages[statusCode].detail,
+        },
+      ],
+    });
+  }
+});
+
+app.get(
+  "/grants",
+  awaitErrorHandlerFactory(
+    async (request: Request, response: Response, next: NextFunction) => {
+      switch (request.version) {
+        case 1: {
+          let grants: Grant[] = await Grant.findAll({ limit: 10 });
+          return response.status(200).json({ data: grants });
+        }
+        default: {
```

Comment:
It's unlikely that we'll ever hit the `default` case here, but I added it because I didn't want the API to fail silently if, somehow, a user got past the version router while still missing a version.

I expect this might come up during the initial development of future versions, and it'll be a good informative error for failing tests.

Reply:
Added a comment in the code to talk about this.